### PR TITLE
basic password implementation

### LINKED ISSUE #81: Obscured input, like `<input type='password'>` on web
Might be an optional component, replacing every character going in to the buffer with `•`.

Obscured input should not be copy or cuttable, but should accept paste.

Option to turn obscuring off via an event would be desirable, like the eye symbols on the web. This suggests a struct such as `pub struct PasswordInput { pub hidden: bool }`.

`editor.get_text()` should return the unobscured string, so the buffer must hold the actual text, only drawing an amount of • equal to the length of the buffer.

Hopefully displaying and holding different values doesn't mess up the click functions.

+ Default font should include the bullet point glyph to work out-of-the-box

### PR DESCRIPTION
**Pull request by bytemunch**
closes #81 

adds bullet (U+2022) to font subset

replaces rendered text with bullets in password fields. could be a cheaper (less overhead) approach to placeholders too.

 - [x] selection in password widgets is buggy, i think because of xoffsets not matching between typed glyphs and replacement bullets.
 - [x] Prevent clipboard read access

### CODE REVIEW

**Comment by Dimchikkk** [comment]
Text in login example looks too tiny on Retina display. Can we have "default" scale factor to be `let scale_factor = primary_window.scale_factor() as f32;` or it's not possible I guess?

**Comment by bytemunch** [comment]
Hacky but we could set the default to something stupid like `0.69696969` and have a system checking if that scale is set on anything, then set the scale to window scale? I'll open an issue

File: src/render.rs [hunk 1]
```diff
@@ -62,26 +63,38 @@ pub(crate) fn cosmic_edit_redraw_buffer(
...
+        if !cosmic_editor.0.buffer().redraw() {
             continue;
         }
 
+        let current_text = cosmic_editor.get_text();
+
+        // intercept text for password inputs
+        if password_opt.is_some() && !current_text.is_empty() {
+            cosmic_editor.set_text(
+                crate::CosmicText::OneStyle("•".repeat(current_text.len())),
```
Comment (Dimchikkk) [inline src/render.rs]: what do you think about making it configurable? Password component will keep "•" char to replace chars to.

Comment (bytemunch) [reply src/render.rs]: Yeah sounds good, I like the idea of a password field that blocks with emoji and this would allow it

**Comment by Dimchikkk** [comment]
🤫 :)

**Comment by bytemunch** [comment]
things are going

https://github.com/StaffEngineer/bevy_cosmic_edit/assets/43527203/894d7af7-d995-4e59-a3d4-60f18eb4fdd5

not well

normal character blocking with the dash is working nicely though! Gotta fix the stuck cursor bug again for password inputs,  they retain a different cursor on deselection.

Not sure why unicode characters make the buffer repeat itself to infinity.

Gonna look in to moving the buffer swapping into their own systems, called before and after the systems that need them.

**Comment by Dimchikkk** [comment]
> Not sure why unicode characters make the buffer repeat itself to infinity.

not good :)

**Comment by bytemunch** [comment]
The `Cursor` `index` field counts UTF-8 single length chars, so moving a cursor / selection between buffers needs to account for this. Will work fine with a bit of number crunching, but only if the characters input are all single length, breaking multi-byte chars in password fields.

Next I'll look into char counting and setting cursor position from that rather than copying and multiplying/dividing the index, as that approach is giving me non-char-boundary issues

**Comment by bytemunch** [comment]
Getting the glyph index from the cursor index seems a bit more complicated than I expected. Think I need to familiarise myself with `LayoutCursor` and how it interacts with `Cursor`.

**Comment by Dimchikkk** [comment]
> Think I need to familiarise myself with `LayoutCursor` and how it interacts with `Cursor`.

Cosmic-text rabbit hole.

**Comment by bytemunch** [comment]
This is notes of things to try for myself :)

Track byte-length difference per character in a hashmap and lookup cursor position to determine offset.

Keep actual text input in a separate buffer like Placeholder. Still need to do byte-length stuff to pass down selection and cursor position accurately.

Find a way in cosmic-text of getting glyph index independent of byte-length.

Track password value independently from buffer, leave the buffer as only blockers. Custom selection, deletion, cursor move events needed.

Ask upstream for glyph index help.

Block multi-byte chars from password input + open a new issue to solve later.

Read up on the hit detection, see how it works and how that determines where to put the cursor. Might shed some light on getting the glyph location.

**Comment by bytemunch** [comment]
After [a](https://security.stackexchange.com/questions/177708/is-it-a-good-idea-to-use-the-entire-unicode-range-to-generate-a-random-password) [lot](https://security.stackexchange.com/questions/85663/is-it-a-good-idea-to-use-non-latin-unicode-characters-in-a-password) of [reading](https://unicode.org/reports/tr15/) I'm leaning more and more on the side of banning multi-byte chars from password inputs. There seems to be good reason (unicode non-determinism) and precedent to do so, but it doesn't feel very forward-thinking or inclusive. Why shouldn't diacritics be allowed in passwords? Makes sense in english sure, but almost every other language needs to mangle words.

Then again, people *should* be using password managers, which are largely ASCII friendly. But we all know they don't.

Lots of people say that many backends don't support unicode, but it's the future now and I'm pretty sure it's not a backend problem anymore.

A counterpoint is that maybe the input is used for something other than a password which does require Unicode support.

@StaffEngineer thoughts on just dropping Unicode in password fields?

**Comment by Dimchikkk** [comment]
@bytemunch yeah, let's drop Unicode in password fields for now :)

**Comment by Dimchikkk** [comment]
Emoji password.... want could be better... so cool 😸

**Comment by Dimchikkk** [comment]
> A counterpoint is that maybe the input is used for something other than a password which does require Unicode support.

Yeah, you could kinda hide information using password field and then reveal it :) aka ANKI cards... but no problem... it can be done in other ways too.

**Comment by Dimchikkk** [comment]
Screenshot 2023-10-27 at 13 15 41

Tha application window/text is so tiny on Retina display... it's good size on my desktop display though :)

**Comment by Dimchikkk** [comment]
Pushed one commit to improve window sizing on Retina.

@bytemunch I experience weird stuff... sometimes when I run app it shows hidden password in output field instead of actual value... but it's hard to reproduce.

**Comment by Dimchikkk** [comment]
image

reproduced ^^ it's very rarely but happens

**Comment by bytemunch** [comment]
Something schedule related, I'll look into the system ordering

**Comment by bytemunch** [comment]
Latest commit fixes the bug far as I can tell, ~20 runs. I'm going to open an issue regarding using custom SystemSets to allow userland fns to depend on internal systems being done

**Review by Dimchikkk — APPROVED** [review]

**Comment by Dimchikkk** [comment]
🎉 🎉 🎉

**Pull request merged**